File: acwing/蓝桥杯每日一题2025/week3/5540.最大限度地提高生产力/5540.最大限度地提高生产力.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <algorithm>

using namespace std;

const int N = 2 * 1e5 + 10;

int c[N];

int main()
{
    int n, q;
    cin >> n >> q;

    for (int i = 0; i < n; i++)
    {
        cin >> c[i];
    }

    for (int i = 0; i < n; i++)
    {
        int t;
        cin >> t;

        c[i] = c[i] - t;
    }

    sort(c, c + n);

    while (q--)
    {
        int v, s;
        cin >> v >> s;
        int l = 0, r = n - 1;

        while (l < r)
        {
            int mid = (l + r) / 2;
            if (c[mid] > s)
                r = mid;
            else
                l = mid + 1;
        }

        if (c[l] > s && n - l >= v)
            cout << "YES" << endl;
        else
            cout << "NO" << endl;
    }

    return 0;
}

File: acwing/蓝桥杯每日一题2025/week3/5540.最大限度地提高生产力/5540.最大限度地提高生产力.py
N = 2 * int(1e5) + 10

n, q = map(int, input().strip().split())

c = list(map(int, input().strip().split()))

t = list(map(int, input().strip().split()))

a = [c[i] - t[i] for i in range(n)]

a.sort()


for _ in range(q):
    v, s = map(int, input().strip().split())
    l = 0
    r = n - 1
    while l < r:
        mid = (l + r + 1) // 2
        if (a[mid] <= s):
            l = mid
        else:
            r = mid - 1

    if a[l] <= s and n - l - 1 >= v:
        print("YES")
    else:
        print("NO")
